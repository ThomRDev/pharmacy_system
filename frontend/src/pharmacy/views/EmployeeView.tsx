import { Box } from "@mui/system"
import { Autocomplete, Button, Grid, InputAdornment, TextField, Typography } from "@mui/material"
import SaveIcon from '@mui/icons-material/Save';
import Paper from '@mui/material/Paper';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TablePagination from '@mui/material/TablePagination';
import TableRow from '@mui/material/TableRow';
import * as React from 'react';
import { getAllEmployees } from "../helpers/employee";
import moment from "moment";
import { useForm } from "react-hook-form";
import { useSelector } from "react-redux";
import { RootState, useAppDispatch } from "../../store";
import { startCreatingNewEmployee } from "../../store/employee/thunks";
import Snackbar from "@mui/material/Snackbar";
import { Alert } from "@mui/lab";
interface Column {
  id: 'id' | 'createdAt' | 'updatedAt' | 'firstname' | 'lastname' | 'email' | 'age' | 'dni' | 'username' | 'role';
  label: string;
  minWidth?: number;
  align?: 'left' | 'right';
  format?: (value: string) => string;
}

const columns: readonly Column[] = [
  { id: 'id', label: 'Identificador', minWidth: 170 },
  { id: 'firstname', align: 'left', label: 'FirstName', minWidth: 100 },
  { id: 'lastname', align: 'left', label: 'LastName', minWidth: 100 },
  { id: 'email', align: 'left', label: 'Email', minWidth: 100 },
  { id: 'age', align: 'left', label: 'Age', minWidth: 100 },
  { id: 'dni', align: 'left', label: 'DNI', minWidth: 100 },
  { id: 'username', align: 'left', label: 'Username', minWidth: 100 },
  {
    id: 'createdAt',
    label: 'CreatedAt',
    minWidth: 170,
    align: 'left',
    format: (value: string) => moment(value).format('MMMM Do YYYY, h:mm:ss a')
  },
  {
    id: 'updatedAt',
    label: 'UpdatedAt',
    minWidth: 170,
    align: 'left',
    format: (value: string) => moment(value).format('MMMM Do YYYY, h:mm:ss a')
  }
];

interface Data {
  id: string;
  createdAt: string;
  updatedAt: string;
  firstname: string;
  lastname: string;
  email: string;
  age: number;
  dni: string;
  username: string;
  role: string;
}

type EmployeeTypes = {
  firstname: string;
  lastname: string;
  email: string;
  age: number;
  dni: string;
  username: string;
  password: string
}

export const EmployeeView = () => {

  const [rows, setRows] = React.useState<Data[]>([])
  const [refetch, setRefecth] = React.useState<null | undefined | boolean>(null)
  const [openNotification, setOpenNotification] = React.useState(false);
  const [page, setPage] = React.useState(0);
  const [rowsPerPage, setRowsPerPage] = React.useState(10);
  const { control, register, handleSubmit, formState: { errors },reset } = useForm<EmployeeTypes>()

  const { messageSaved, error } = useSelector((state: RootState) => state.employee)

  const dispatch = useAppDispatch()
  const handleCloseNotification = (event?: React.SyntheticEvent | Event, reason?: string) => {
    if (reason === 'clickaway') {
      return;
    }

    setOpenNotification(false);
  };
  React.useEffect(() => {
    if (!!messageSaved) {
      setRefecth(!refetch)
    }
    // if(!!messageSaved) {
    // }
  }, [messageSaved])

  React.useEffect(() => {
    getAllEmployees().then((employees) => {
      // console.log();
      setRows(employees.data)
    })
  }, [refetch])

  const handleChangePage = (event: unknown, newPage: number) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (event: React.ChangeEvent<HTMLInputElement>) => {
    setRowsPerPage(+event.target.value);
    setPage(0);
  };

  const onCreateEmployee = (data: any) => {
    console.log(data);
    dispatch(startCreatingNewEmployee({
      ...data,
      age: +data.age,
      role: 'ADMIN'
    }))
    setOpenNotification(true)
    reset()
  }

  return (
    <>
      <Typography variant="h5" sx={{ mb: "1.5em" }} >Employees</Typography>
      <Box sx={{ backgroundColor: "#2D2F2E", borderRadius: "5px", p: "1em", width: "98%", mb: "1em" }}>
        <form style={{ display: "flex", flexDirection: "column", gap: "1em" }} onSubmit={handleSubmit(onCreateEmployee)} >
          <Grid container sx={{ gap: "1em", flexWrap: "wrap" }}>
            <TextField
              aria-label="txtFirstName"
              id="txtFirstName"
              label="First Name"
              variant="filled"
              autoComplete="false"
              {...register("firstname", {
                required: {
                  value: true,
                  message: "El FirstName es necesario"
                }
              })}
              helperText={errors.firstname?.message}
              error={!!errors.firstname}
            />
            <TextField
              aria-label="txtLastName"
              id="txtLastName"
              label="Last Name"
              variant="filled"
              autoComplete="false"
              {...register("lastname", {
                required: {
                  value: true,
                  message: "El Lastname es necesario"
                }
              })}
              helperText={errors.lastname?.message}
              error={!!errors.lastname}
            />
            <TextField
              aria-label="txtEmail"
              id="txtEmail"
              label="Email"
              variant="filled"
              autoComplete="false"
              {...register("email", {
                required: "El email es necesario",
                pattern: {
                  value: /\S+@\S+\.\S+/,
                  message: "No es un Email valido"
                }
              })}
              type="email"
              helperText={errors.email?.message}
              error={!!errors.email}
            />
            <TextField
              aria-label="txtEdad"
              id="txtEdad"
              label="Edad"
              variant="filled"
              autoComplete="false"
              type={'number'}
              {...register("age", {
                required: "La edad es necesaria",
                min: {
                  value: 18,
                  message: 'El empleado debe ser mayor a 17'
                },
                max: {
                  value: 70,
                  message: 'El empleado debe ser menor a 70'
                }
              })}
              helperText={errors.age?.message}
              error={!!errors.age}
            />
            <TextField
              aria-label="txtDni"
              id="txtDni"
              label="DNI"
              variant="filled"
              autoComplete="false"
              {...register("dni", {
                required: "El dni es necesario",
                minLength: {
                  value: 8,
                  message: 'Debe de tener 8 caracteres'
                },
                maxLength: {
                  value: 8,
                  message: 'Debe de tener 8 caracteres'
                }
              })}
              helperText={errors.dni?.message}
              error={!!errors.dni}
            />
            <TextField
              aria-label="txtUsername"
              id="txtUsername"
              label="Username"
              variant="filled"
              autoComplete="false"
              {...register("username", {
                required: {
                  value: true,
                  message: "El username es necesario"
                },
                minLength: {
                  value: 5,
                  message: "El username debe de tener 5 caracteres como mínimo"
                },
              })}
              helperText={errors.username?.message}
              error={!!errors.username}
            />
            <TextField
              aria-label="txtPassword"
              id="txtPassword"
              label="Password"
              variant="filled"
              autoComplete="false"
              type="password"
              inputProps={{
                "aria-label": 'password'
              }}
              placeholder="***********"

              {...register("password", {
                required: {
                  value: true,
                  message: "El password es necesario"
                },
                minLength: {
                  value: 5,
                  message: "El password debe de tener 5 caracteres como mínimo"
                },
              })}
              error={!!errors.password}
              helperText={errors.password?.message}
            />
          </Grid>
          <Grid container>
            <Button
              aria-label="btn-save"
              id="btn-save"
              type="submit" variant="contained" sx={{ background: "#068E9E", "&:hover": { background: "#068f9eac" } }}><SaveIcon />&nbsp;Guardar empleado</Button>
          </Grid>
        </form>
      </Box>
      <Paper sx={{ width: '95%', overflow: 'hidden' }}>
        <TableContainer sx={{ maxHeight: 440 }}>
          <Table stickyHeader aria-label="sticky table">
            <TableHead>
              <TableRow>
                {columns.map((column) => (
                  <TableCell
                    key={column.id}
                    align={column.align}
                    style={{ minWidth: column.minWidth }}
                  >
                    {column.label}
                  </TableCell>
                ))}
              </TableRow>
            </TableHead>
            <TableBody>
              {rows
                .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
                .map((row) => {
                  return (
                    <TableRow hover role="checkbox" tabIndex={-1} key={row.id}>
                      {columns.map((column) => {
                        const value = row[column.id];
                        return (
                          <TableCell key={column.id} align={column.align}>
                            {column.format && typeof value === 'string'
                              ? column.format(value)
                              : value}
                          </TableCell>
                        );
                      })}
                    </TableRow>
                  );
                })}
            </TableBody>
          </Table>
        </TableContainer>
        <TablePagination
          rowsPerPageOptions={[10, 25, 100]}
          component="div"
          count={rows.length}
          rowsPerPage={rowsPerPage}
          page={page}
          onPageChange={handleChangePage}
          onRowsPerPageChange={handleChangeRowsPerPage}
        />
      </Paper>
      {
        !!messageSaved && <Snackbar open={openNotification} autoHideDuration={6000} onClose={handleCloseNotification}>
          <Alert onClose={handleCloseNotification} severity="success" sx={{ width: '100%' }}>{messageSaved}</Alert>
        </Snackbar>
      }
      {
        !!error && <Snackbar open={openNotification} autoHideDuration={6000} onClose={handleCloseNotification}>
          <Alert onClose={handleCloseNotification} severity="error" sx={{ width: '100%' }}>{error}</Alert>
        </Snackbar>
      }
    </>
  )
}
