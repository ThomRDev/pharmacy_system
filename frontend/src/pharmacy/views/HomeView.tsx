import React, { useState, useEffect, useRef } from 'react'
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
  BarElement,
  ChartData
} from 'chart.js';

import { Line ,Bar } from 'react-chartjs-2';
import { getAllProducts, getMostSelledProducts } from '../helpers/product';
import { Box, Grid, Typography } from '@mui/material';
import { Resume } from '../components/Resume';

ChartJS.register(
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
  BarElement
);

export const HomeView = () => {
  const [chartRe, setChartRe] = useState<any>(null)
  const [ chartMost,setChartMost] = useState<any>(null)

  const ref = useRef(true)

  useEffect(()=>{

    ref.current = true
    
    setChartRe(null)
    getAllProducts()
    .then(products=>{
      if(ref.current){
        setChartRe(products.data)
      }
    })

    getMostSelledProducts()
    .then(products=>{
      if(ref.current){
        setChartMost(products.data)
      }
    })

    return () => {
      ref.current = false
    }
  },[])

  var dataReg: ChartData<"bar", any, unknown> = {
    labels: chartRe?.map((product:any) => product.name),
    
    datasets: [{
      label: `Productos Registrados`,
      
      data: chartRe?.map((product:any) => product.stock),
      backgroundColor: [
        'rgba(255, 99, 132, 0.2)',
        'rgba(54, 162, 235, 0.2)',
        'rgba(255, 206, 86, 0.2)',
        'rgba(75, 192, 192, 0.2)',
        'rgba(153, 102, 255, 0.2)',
        'rgba(255, 159, 64, 0.2)'
      ],
      borderColor: [
        'rgba(255, 99, 132, 1)',
        'rgba(54, 162, 235, 1)',
        'rgba(255, 206, 86, 1)',
        'rgba(75, 192, 192, 1)',
        'rgba(153, 102, 255, 1)',
        'rgba(255, 159, 64, 1)'
      ],
      borderWidth: 1
    }]
  }
  var dataMost = {
    labels: chartMost?.map((product:any) => product.name),
    datasets: [{
      label: `Vendido`,
      data: chartMost?.map((product:any) => product.vendido),
      backgroundColor: [
        'rgba(255, 99, 132, 0.2)',
        'rgba(54, 162, 235, 0.2)',
        'rgba(255, 206, 86, 0.2)',
        'rgba(75, 192, 192, 0.2)',
        'rgba(153, 102, 255, 0.2)',
        'rgba(255, 159, 64, 0.2)'
      ],
      borderColor: [
        'rgba(255, 99, 132, 1)',
        'rgba(54, 162, 235, 1)',
        'rgba(255, 206, 86, 1)',
        'rgba(75, 192, 192, 1)',
        'rgba(153, 102, 255, 1)',
        'rgba(255, 159, 64, 1)'
      ],
      borderWidth: 1
    }]
  }

  var options = {
    maintainAspectRatio: false,
    scales: {
    },
    legend: {
      labels: {
        fontSize: 25,
      },
    },
  }

  return (
    <Grid container sx={{ width:"100%" }}>
      <Resume />
      <Grid container justifyContent={"center"} sx={{ gap:3 }} >
        <Box sx={{
          width: "600px",
          height: "600px",
          // boxShadow: "0px 5px 5px -3px rgb(0 0 0 / 20%), 0px 8px 10px 1px rgb(0 0 0 / 14%), 0px 3px 14px 2px rgb(0 0 0 / 12)",
          border:"4px solid rgba(255,255,255,0.1)",
          borderRadius:"10px",
          overflow: "hidden",
          display:"flex",
          flexDirection:"column"
        }}>
          <Box sx={{backgroundColor :"#1C1F1E",padding :"1em",textAlign:"center" }}>
            <Typography variant='h5'>Productos Registrados</Typography>
          </Box>
          <Box sx={{ flex:"1",padding :"2em" }}>
            <Bar
              data={dataReg}
              height={100}
              options={options}
              aria-label='canvas-registrados'
            />
          </Box>
        </Box>
        <Box sx={{
          width: "600px",
          height: "600px",
          // boxShadow: "0px 5px 5px -3px rgb(0 0 0 / 20%), 0px 8px 10px 1px rgb(0 0 0 / 14%), 0px 3px 14px 2px rgb(0 0 0 / 12)",
          border:"4px solid rgba(255,255,255,0.1)",
          borderRadius:"10px",
          overflow: "hidden",
          display:"flex",
          flexDirection:"column"
        }}>
          <Box sx={{backgroundColor :"#1C1F1E",padding :"1em",textAlign:"center" }}>
            <Typography variant='h5'>Productos mas vendidos</Typography>
          </Box>
          <Box sx={{ flex:"1",padding :"2em" }}>
            <Line
              data={dataMost}
              height={100}
              
              options={options}
              aria-label='canvas-mas-vendido'
            />
          </Box>
        </Box>
      </Grid>
    </Grid>
  )
}
