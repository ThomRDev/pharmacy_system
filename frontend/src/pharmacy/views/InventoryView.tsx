import { useGetProductsQuery } from "../../store/apis/invetoryApi"
import { GridColDef, GridValueGetterParams } from '@mui/x-data-grid';
import { Alert, Backdrop, Box, Button, Fade, Grid, IconButton, Modal , Snackbar, Typography } from "@mui/material";
import moment from "moment"
import { EditOutlined } from "@mui/icons-material";
import AddCircleIcon from '@mui/icons-material/AddCircle';

import StoreIcon from '@mui/icons-material/Store';

import { Loader } from "../components/Loader";
import { ModalEditProduct } from "../components/ModalEditProduct";
import { modal_style } from "../utils";
import { InvetoryTable } from "../components/InvetoryTable";
import { useState } from "react";
import { ModalCreateProduct } from "../components/ModalCreateProduct";
import { RootState, useAppDispatch } from "../../store";
import { useSelector } from "react-redux";
import { clearProduct } from "../../store/inventory/inventorySlice";

export const InventoryView = () => {
  const [openNotification, setOpenNotification] = useState(false);
  const { messageSaved,isSaving,error } = useSelector((state:RootState)=>state.product)

  const dispatch = useAppDispatch()

  const handleCloseNotification = (event?: React.SyntheticEvent | Event, reason?: string) => {
    if (reason === 'clickaway') {
      return;
    }

    setOpenNotification(false);
  };

  const handleClickNotification = () => {
    setOpenNotification(true);
  };

  const productsResponse = useGetProductsQuery({})
  const [productId,setProductId] = useState<string | null>(null)
  const [modalActive,setModalActive] = useState({
    edit : false,
    create:false
  })

  const [openModal, setOpenModal] = useState(false);
  const onOpenModal = () => {
    dispatch(clearProduct())
    setOpenModal(true)
  }
  const onCloseModal = () => {
    productsResponse.refetch()
    setOpenModal(false)
  };


  const columns: GridColDef[] = [
    { field: 'name', headerName: 'Producto', width: 110, headerAlign: "center", align: "center" },
    {
      field: 'edit', headerName: 'Editar',
      sortable: false, width: 80, headerAlign: "center", align: "center", renderCell: (params) => {
        return <IconButton sx={{
          background: "#097CEA",
          "&:hover": {
            background: "#097deaa9",
          }
        }}
          onClick={() => {
            setModalActive((modal:any)=>({ edit:true,create:false }))
            setProductId(params.row?.id)
            onOpenModal()
          }
          }
        > <EditOutlined sx={{ color: "white" }} /></IconButton>
      }
    },
    { field: 'therapeutic_description', headerName: 'Desc. Terapeutica', width: 230, headerAlign: "center", align: "center" },
    {
      field: 'price',
      headerName: 'Precio',
      type: 'number',
      width: 100,
      headerAlign: "center", align: "center",
      valueGetter: (params: GridValueGetterParams) => {
        return `s/ ${params.value}.00`;
      }
    },
    {
      field: 'stock',
      headerName: 'Stock',
      width: 80,
      headerAlign: "center", align: "center",
      valueGetter: (params: GridValueGetterParams) => {
        return `${params.value} u`;
      }
    },
    {
      field: "createdAt", headerName: "Fecha de creaciÃ³n", width: 210, headerAlign: "center", align: "center",
      valueGetter: (params: GridValueGetterParams) => {
        return moment(params.value).format('MMMM Do YYYY, h:mm:ss a');
      }
    },
    {
      field: "expiration_date", headerName: "Fecha de expiracion", width: 210, headerAlign: "center", align: "center",
      valueGetter: (params: GridValueGetterParams) => {
        return moment(params.value).format('MMMM Do YYYY, h:mm:ss a');
      }
    },
    {
      field: "updatedAt", headerName: "Fecha de actualizacion", width: 210, headerAlign: "center", align: "center",
      valueGetter: (params: GridValueGetterParams) => {
        return moment(params.value).format('MMMM Do YYYY, h:mm:ss a');
      }
    },
    {
      field: "laboratory", headerName: "Laboratorio", width: 180, headerAlign: "center", align: "center",
      valueGetter: (params: GridValueGetterParams) => {
        return `${params.value?.name}`
      }
    }
  ];
  return (
    <div style={{ height: 400, width: '100%' }}>
      <Typography variant="h3" mb={3} sx={{ display: "flex", alignItems: "center" }}><StoreIcon sx={{ fontSize: "1.5em" }} />&nbsp;Inventario de Productos</Typography>
      { productsResponse.isLoading ? <Loader /> : <>
        <Grid container mb={2} >
          <Button
          aria-label="btn-add-product" 
          id="btn-add-product"
          onClick={()=>{
            onOpenModal()
            setModalActive((modal:any)=>({ create:true,edit:false }))
          }}
          variant="contained" sx={{ background: "#068E9E", "&:hover": { background: "#068f9eac" } }}><AddCircleIcon />&nbsp;Agregar Nuevo producto</Button>
        </Grid>
        <InvetoryTable rows={productsResponse.data?.data} columns={columns} />
      </> }
      <Modal
        aria-labelledby="transition-modal-title"
        aria-describedby="transition-modal-description"
        open={openModal}
        onClose={onCloseModal}
        sx={{ background: "rgba(0,0,0,0.8)" }}
        closeAfterTransition
        BackdropComponent={Backdrop}
        BackdropProps={{
          timeout: 500,
        }}
        
      >
        <Fade in={openModal}>
          <Box sx={modal_style}>
            { modalActive.edit && <ModalEditProduct productId={productId as string} onCloseModal={onCloseModal} handleClickNotification={handleClickNotification} /> }
            { modalActive.create && <ModalCreateProduct onCloseModal={onCloseModal} handleClickNotification={handleClickNotification} /> }
            {/* // <ModalEditProduct {...productResponse.data?.data} /> } */}
          </Box>
        </Fade>
      </Modal>

      {
        !!messageSaved &&  <Snackbar open={openNotification} autoHideDuration={6000} onClose={handleCloseNotification}>
          <Alert onClose={handleCloseNotification} severity="success" sx={{ width: '100%' }}>{ messageSaved }</Alert>
        </Snackbar>
      }
      {
        !!error &&  <Snackbar open={openNotification} autoHideDuration={6000} onClose={handleCloseNotification}>
        <Alert onClose={handleCloseNotification} severity="error" sx={{ width: '100%' }}>{ error }</Alert>
      </Snackbar>
      }
    </div>
  )
}
