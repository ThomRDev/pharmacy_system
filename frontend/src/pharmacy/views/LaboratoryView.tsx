import {
  Alert,
  Backdrop,
  Box,
  Button,
  Fade,
  Grid,
  IconButton,
  Modal,
  Snackbar,
  Typography,
} from "@mui/material";
import BiotechIcon from "@mui/icons-material/Biotech";
import AddIcon from "@mui/icons-material/Add";
import { LaboratoryTable } from "../components/LaboratoryTable";
import { EditOutlined } from "@mui/icons-material";
import { GridColDef, GridValueGetterParams } from "@mui/x-data-grid";
import moment from "moment";
import { useGetLaboratoriesQuery } from "../../store/apis/laboratoryApi";
import { useEffect, useState } from "react";
import { ModalCreateLaboratory } from "../components/ModalCreateLaboratory";
import { Loader } from "../components/Loader";
import { useSelector } from "react-redux";
import { RootState } from "../../store";
import { ModalEditLaboratory } from "../components/ModalEditLaboratory";


import { getLaboratories } from './../helpers/laboratory'

export const LaboratoryView = () => {
  const laboratoryResponse = useGetLaboratoriesQuery({});
  const [openNotification, setOpenNotification] = useState(false);
  const { messageSaved,error } = useSelector((state:RootState)=>state.laboratoryStore)
  const [laboratoryId, setLaboratoryId] = useState<string | null>(null);
  const [modalActive, setModalActive] = useState({
    edit: false,
    create: false,
  });

  const [laboratories,setLaboratories] = useState<any>(null)

  const [refetch,setRefetch] = useState<boolean>(false)

  useEffect(()=>{
    getLaboratories()
    .then(response=>{
      setLaboratories(response.data)
    })
  },[refetch])

  const [openModal, setOpenModal] = useState(false);
  const onOpenModal = () => {
    setOpenModal(true);
  };
  const onCloseModal = () => {
    // laboratoryResponse.refetch();
    setRefetch(f=>!f)
    console.log('refetch');
    setOpenModal(false);
  };

  const handleClickNotification = () => {
    setOpenNotification(true);
  };
  const handleCloseNotification = (event?: React.SyntheticEvent | Event, reason?: string) => {
    if (reason === 'clickaway') {
      return;
    }

    setOpenNotification(false);
  };

  const columns: GridColDef[] = [
    {
      field: "name",
      headerName: "Nombre",
      width: 250,
      headerAlign: "center",
      align: "center",
    },
    {
      field: "edit",
      headerName: "Editar",
      sortable: false,
      width: 100,
      headerAlign: "center",
      align: "center",
      renderCell: (params) => {
        return (
          <IconButton
            sx={{
              background: "#097CEA",
              "&:hover": {
                background: "#097deaa9",
              },
            }}
            onClick={() => {
              setModalActive((modal:any)=>({ edit:true,create:false }))
              setLaboratoryId(params.row?.id)
              onOpenModal()
            }}
          >
            {" "}
            <EditOutlined sx={{ color: "white" }} />
          </IconButton>
        );
      },
    },
    {
      field: "description",
      headerName: "Descripcion",
      width: 300,
      headerAlign: "center",
      align: "center",
    },
    {
      field: "address",
      headerName: "Direccion",
      width: 300,
      headerAlign: "center",
      align: "center",
    },
    {
      field: "createdAt",
      headerName: "Fecha de creaciÃ³n",
      width: 210,
      headerAlign: "center",
      align: "center",
      valueGetter: (params: GridValueGetterParams) => {
        return moment(params.value).format("MMMM Do YYYY, h:mm:ss a");
      },
    },
    {
      field: "updatedAt",
      headerName: "Fecha de actualizacion",
      width: 210,
      headerAlign: "center",
      align: "center",
      valueGetter: (params: GridValueGetterParams) => {
        return moment(params.value).format("MMMM Do YYYY, h:mm:ss a");
      },
    },
  ];
  return (
    <div style={{ height: 400, width: "100%" }}>
      <Typography
        variant="h3"
        mb={3}
        sx={{ display: "flex", alignItems: "center" }}
      >
        <BiotechIcon sx={{ fontSize: "1.5em" }} />
        &nbsp;Laboratorios
      </Typography>
      { laboratoryResponse.isLoading ? <Loader /> : 
      <>
        <Grid container mb={2}>
          <Button
            variant="contained"
            sx={{
              background: "#068E9E",
              "&:hover": { background: "#068f9eac" },
            }}
            onClick={()=>{
              onOpenModal()
              setModalActive((modal:any)=>({ create:true,edit:false }))
            }}
          >
            <AddIcon />
            &nbsp;Agregar un nuevo laboratorio
          </Button>
        </Grid>
        <LaboratoryTable rows={laboratoryResponse.data?.data} columns={columns} />
      </>}
      <Modal
        aria-labelledby="transition-modal-title"
        aria-describedby="transition-modal-description"
        open={openModal}
        onClose={onCloseModal}
        sx={{ background: "rgba(0,0,0,0.8)" }}
        closeAfterTransition
        BackdropComponent={Backdrop}
        BackdropProps={{
          timeout: 500,
        }}
      >
        <Fade in={openModal}>
          <Box
            sx={{
              position: "absolute" as "absolute",
              top: "50%",
              left: "50%",
              transform: "translate(-50%, -50%)",
              width: "900px",
              bgcolor: "background.paper",
              border: "2px solid #000",
              boxShadow: 24,
              p: 4,
              background: "#17191A",
            }}
          >
            {
              modalActive.edit && 
              <ModalEditLaboratory 
                laboratoryId={laboratoryId as string} 
                onCloseModal={onCloseModal} 
                handleClickNotification={handleClickNotification} 
              />
            }
            {
              modalActive.create && 
              <ModalCreateLaboratory 
                onCloseModal={onCloseModal} 
                handleClickNotification={handleClickNotification} 
              />
            }
          </Box>
        </Fade>
      </Modal>
      {
        !!messageSaved &&  <Snackbar open={openNotification} autoHideDuration={6000} onClose={handleCloseNotification}>
          <Alert onClose={handleCloseNotification} severity="success" sx={{ width: '100%' }}>{ messageSaved }</Alert>
        </Snackbar>
      }
      {
        !!error &&  <Snackbar open={openNotification} autoHideDuration={6000} onClose={handleCloseNotification}>
        <Alert onClose={handleCloseNotification} severity="error" sx={{ width: '100%' }}>{ error }</Alert>
      </Snackbar>
      }
    </div>
  );
};
