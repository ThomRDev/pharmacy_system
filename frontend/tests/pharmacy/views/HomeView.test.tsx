import { expect } from '@jest/globals';
import { render, screen, waitFor } from '@testing-library/react';
import "jest";
import { HomeView } from '../../../src/pharmacy/views/HomeView';
import { mostSelled, products } from '../../fixtures/homeFixtures';

const mockGetAllProducts = jest.fn();
const mockGetMostSelledProducts = jest.fn();


jest.mock("../../../src/pharmacy/helpers/product",()=>({
  getMostSelledProducts : () => mockGetMostSelledProducts(),
  getAllProducts : () => mockGetAllProducts()
}))

window.ResizeObserver =
    window.ResizeObserver ||
    jest.fn().mockImplementation(() => ({
        disconnect: jest.fn(),
        observe: jest.fn(),
        unobserve: jest.fn(),
    }));

describe('Testing al componente de <HomeView />', () => {
  test('Deberia renderizar hacer match como el snapshot', async () => {
    mockGetMostSelledProducts.mockResolvedValue({
      data : mostSelled
    })
    mockGetAllProducts.mockResolvedValue({
      data : products
    })
    const { container } = render(
        <HomeView />
    )
    await waitFor(() =>{
      expect(container).toMatchSnapshot();
    })
  })
  test('Deberia renderizar el canvas para el grafico de los productos mas vendidos', async () => {
    mockGetMostSelledProducts.mockResolvedValue({
      data : mostSelled
    })
    mockGetAllProducts.mockResolvedValue({
      data : products
    })
    render(
      <HomeView />
    )
    await waitFor(() =>{
      expect(screen.getByText('Productos mas vendidos')).toBeTruthy()
    })
    await waitFor(() =>{
      expect(screen.getByLabelText('canvas-mas-vendido')).toBeTruthy()
    })
  })
  test('Deberia renderizar el canvas para el grafico de todos los productos registrados', async () => {
    mockGetMostSelledProducts.mockResolvedValue({
      data : mostSelled
    })
    mockGetAllProducts.mockResolvedValue({
      data : products
    })
    render(
      <HomeView />
    )
    await waitFor(() =>{
      expect(screen.getByText('Productos Registrados')).toBeTruthy()
    })
    await waitFor(() =>{
      expect(screen.getByLabelText('canvas-registrados')).toBeTruthy()
    })
  })
})