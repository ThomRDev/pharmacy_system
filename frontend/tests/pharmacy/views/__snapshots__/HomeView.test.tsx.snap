// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing al componente de <HomeView /> Deberia renderizar hacer match como el snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container css-t78dm4-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-1bbjngc-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-p79t05"
      >
        <div
          class="MuiBox-root css-7cxx3r"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            Productos Registrados
          </h5>
        </div>
        <div
          class="MuiBox-root css-uozptx"
        >
          <canvas
            aria-label="canvas-registrados"
            height="0"
            role="img"
            style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
            width="0"
          />
        </div>
      </div>
      <div
        class="MuiBox-root css-p79t05"
      >
        <div
          class="MuiBox-root css-7cxx3r"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            Productos mas vendidos
          </h5>
        </div>
        <div
          class="MuiBox-root css-uozptx"
        >
          <canvas
            aria-label="canvas-mas-vendido"
            height="0"
            role="img"
            style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
            width="0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
